### PR TITLE
Improving performance of nudging routines

### PR DESCRIPTION
**Pull request by afanfa**
TYPE: enhancement

KEYWORDS: performance, nudging, better data layout

SOURCE: Alessandro Fanfarillo, NCAR

DESCRIPTION OF CHANGES: The observations used in the nudging routines are organized as Array of Structures. This layout penalizes performance in many ways. This PR reorganize the layout to be a Structure of Arrays.

ISSUE: None

TESTS CONDUCTED: All the tests pass on Cheyenne except reach! According to @jmccreight reach shouldn't even be using nudging. I am creating the PR to test reach using TravisCI.

NOTES: The PR requires some cleanup.

### Checklist
Merging the PR depends on following checklist being completed. Add `X` between each of the square 
brackets if they are completed in the PR itself. If a bullet is not relevant to you, please comment 
on why below the bullet.

 - [ ] Closes issue #xxxx (An issue must exist or be created to be closed. The issue describes and documents the problem and general solution, the PR describes the technical details of the solution.) 
 - [ ] Tests added (unit tests and/or regression/integration tests)
 - [ ] Backwards compatible
 - [ ] Requires new files? If so, how to generate them.
 - [ ] Fully documented
 - [ ] Short description in the Development section of `NEWS.md`

### CODE REVIEW

**Comment by aubreyd** [comment]
Tested new code against master using NWMv2.1 CONUS channel-only configuration. The 4-day simulations produced identical solutions (bit-for-bit CHRTOUT, HYDRO_RST, and nudgingLastObs) with the following runtimes (all using 4 nodes on cheyenne):
- master: 29 minutes
- improved-nudging: 21.5 minutes

**Review by rcabell — DISMISSED** [review]
This looks ok to me, with the exception of one (minor) concern: the new Structure-of-Arrays code has been added but the old Array-of-Structures code has been partially removed, but partially commented-out. If the new code is bit-identical and faster, it should completely replace the older version. I'm ok with merging as-is, but we may want a follow-on PR to clean out the old code.

File: trunk/NDHMS/nudging/module_stream_nudging.F [hunk 1]
```diff
@@ -1136,14 +1178,20 @@ subroutine setup_stream_nudging(hydroDT)
...
       call mpp_land_bcast_int1d_root(obsTimeStr(tt)%obsStr(:)%obsStaticInd, iiImage)
-      call mpp_land_bcast_char1d_root(obsStaticStr(:)%usgsId, iiImage)
-      call mpp_land_bcast_int1d_root(obsStaticStr(:)%obsCellInd, iiImage)
-      call mpp_land_bcast_real_1d_root(obsStaticStr(:)%R, iiImage)
-      call mpp_land_bcast_real_1d_root(obsStaticStr(:)%G, iiImage)
-      call mpp_land_bcast_real_1d_root(obsStaticStr(:)%tau, iiImage)
-
-      nStatic = sum( (/ (1, cc=1,size(obsStaticStr(:)%obsCellInd)) /), &
-                     mask=obsStaticStr(:)%obsCellInd .ne. 0            )
+      call mpp_land_bcast_char1d_root(obsStaticStr_SoA%usgsId(:), iiImage)
```
Comment (scrasmussen) [inline trunk/NDHMS/nudging/module_stream_nudging.F:1181]: This line seems like it should be commented out right? Since it is repeated at line 1187

Comment (rcabell) [reply trunk/NDHMS/nudging/module_stream_nudging.F:1181]: ...or just removed, since it's clearly redundant

Comment (scrasmussen) [reply trunk/NDHMS/nudging/module_stream_nudging.F:1152]: Ok, removed that line and removed a few other commented out code blocks

**Review by scrasmussen — APPROVED** [review]
Looks good to me now!

**Review by rcabell — APPROVED** [review]
@aubreyd's nudging tests show no change in solution or timing with @scrasmussen's latest code cleanup, so I will approve.

**Pull request merged**